src/backwardproblem.cpp

```diff
@@ -59,44 +65,55 @@ void BackwardProblem::workBackwardProblem() {
     }
 
     int it = model->nt() - 1;
-    model->initializeB(xB, dxB, xQB);
-    handleDataPointB(it);
-    solver->setupB(&which, model->getTimepoint(it), model, xB, dxB, xQB);
+    /* If we have posteq, infty timepoints were already treated */
+    for (int jt = model->nt() - 1; jt >= 0; jt--)
+        if (std::isinf(model->getTimepoint(it)))
+            --it;
+    if (it < model->nt() - 1){
+        model->initializeB(xB, dxB, dxB);
+    } else {
+        model->initializeB(xB, dxB, xQB);
+    }
 
-    --it;
+    if ((it >= 0 || discs.size() > 0) && model->getTimepoint(it) > model->t0())
+    {
+        handleDataPointB(it);
+        solver->setupB(&which, model->getTimepoint(it), model, xB, dxB, xQB);
+        --it;
 
-    while (it >= 0 || discs.size() > 0) {
+        while (it >= 0 || discs.size() > 0) {
+            /* check if next timepoint is a discontinuity or a data-point */
+            double tnext = getTnext(it);
 
-        /* check if next timepoint is a discontinuity or a data-point */
-        double tnext = getTnext(it);
+            if (tnext < t) {
+                solver->runB(tnext);
+                solver->writeSolutionB(&t, xB, dxB, xQB, this->which);
+            }
 
-        if (tnext < t) {
-            solver->runB(tnext);
-            solver->writeSolutionB(&t, xB, dxB, xQB, this->which);
-        }
+            /* handle discontinuity */
+            if (tnext > model->getTimepoint(it)) {
+                handleEventB();
+            }
 
-        /* handle discontinuity */
-        if (tnext > model->getTimepoint(it)) {
-            handleEventB();
-        }
+            /* handle data-point */
+            if (tnext == model->getTimepoint(it)) {
+                handleDataPointB(it);
+                it--;
+            }
 
-        /* handle data-point */
-        if (tnext == model->getTimepoint(it)) {
-            handleDataPointB(it);
-            it--;
+            /* reinit states */
+            solver->reInitB(which, t, xB, dxB);
+            solver->quadReInitB(which, xQB);
         }
 
-        /* reinit states */
-        solver->reInitB(which, t, xB, dxB);
-        solver->quadReInitB(which, xQB);
+        /* we still need to integrate from first datapoint to tstart */
```

Comment:
I believe this code should be outside the loop, we only want to run this once.

Reply:
True. Done.